Unrolling of big collection constructors fails to compile for `const` variables

With #1744 we introduced code which unrolls constructors of huge containers. The code we now generate is incorrect for `const` variables which do not live in a scope (it does work for `global`s though which are managed more thightly by the module).

```ruby
module foo;

# Ctrs for collections of more than 10 elements get unrolled.
const CRC16_TABLE: vector<uint16> = [
  0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10
];
```

```console
$ spicyc -dj foo.spicy
foo_198ff91772b68fab-628dbe1417976318.cc:12:84: error: non-local lambda expression cannot have a capture-default
    const ::hilti::rt::Vector<::hilti::rt::integer::safe<uint16_t>> CRC16_TABLE = [&]() { ...
                                                                                   ^
```

We probably need to adjust or even drop the capture for declarations emitted outside of a proper C++ scope. ~We still need to make sure that a constructor can reference other `const` variables though.~ A declaration of a `const` can only refer to other `const` variables; since `const` variables are emitted as non-local variables we do not need to capture them.

--- a/hilti/toolchain/src/compiler/codegen/ctors.cc
+++ b/hilti/toolchain/src/compiler/codegen/ctors.cc
@@ -116,7 +116,13 @@ struct Visitor : hilti::visitor::PreOrder {
                                                     }),
                                     " ");
 
-            result = fmt("[&]() { auto __xs = ::hilti::rt::Map<%s, %s>(); %s return __xs; }()", k, v, elems);
+            // If we are at block scope capture other variables so they can be
+            // used in the ctr. Outside of block scope we are emitting a
+            // non-local `const` (into a namespace) which can reference only
+            // other `const` variables which since they are non-locals as well
+            // can be referenced without capturing.
+            auto captures = (cg->cxxBlock() == nullptr) ? "" : "&";
+            result = fmt("[%s]() { auto __xs = ::hilti::rt::Map<%s, %s>(); %s return __xs; }()", captures, k, v, elems);
         }
 
         else
@@ -199,7 +205,13 @@ struct Visitor : hilti::visitor::PreOrder {
                                            [this](const auto& e) { return fmt("__xs.insert(%s);", cg->compile(e)); }),
                            " ");
 
-            result = fmt("[&]() { auto __xs = ::hilti::rt::Set<%s>(); %s return __xs; }()", k, elems);
+            // If we are at block scope capture other variables so they can be
+            // used in the ctr. Outside of block scope we are emitting a
+            // non-local `const` (into a namespace) which can reference only
+            // other `const` variables which since they are non-locals as well
+            // can be referenced without capturing.
+            auto captures = (cg->cxxBlock() == nullptr) ? "" : "&";
+            result = fmt("[%s]() { auto __xs = ::hilti::rt::Set<%s>(); %s return __xs; }()", captures, k, elems);
         }
 
         else
@@ -287,8 +299,14 @@ struct Visitor : hilti::visitor::PreOrder {
                                                     }),
                                     " ");
 
-            result = fmt("[&]() { auto __xs = ::hilti::rt::Vector<%s%s>(); __xs.reserve(%d); %s return __xs; }()", x,
-                         allocator, size, elems);
+            // If we are at block scope capture other variables so they can be
+            // used in the ctr. Outside of block scope we are emitting a
+            // non-local `const` (into a namespace) which can reference only
+            // other `const` variables which since they are non-locals as well
+            // can be referenced without capturing.
+            auto captures = (cg->cxxBlock() == nullptr) ? "" : "&";
+            result = fmt("[%s]() { auto __xs = ::hilti::rt::Vector<%s%s>(); __xs.reserve(%d); %s return __xs; }()",
+                         captures, x, allocator, size, elems);
         }
 
         else

